### PR TITLE
Update hadoop to 3.2.0

### PR DESCRIPTION
**Pull request by rajatrj20**
When presto runs on Hadoop 3 with `dfs.encrypt.data.transfer=true`, queries fails with error:

> com.facebook.presto.spi.PrestoException: Error committing write to Hive 
> 
> Caused by: org.apache.hadoop.ipc.RemoteException: File /tmp/presto-hadoop/a38381f4-fcc3-431d-bd73-bb0d872a1572/20211106_073946_00003_k9xuf_070809d7-6c9b-4d5f-ac17-86a998d2958d could only be written to 0 of the 1 minReplication nodes. There are 3 datanode(s) running and 3 node(s) are excluded in this operation.

and on HDFS side following error is thrown:
>javax.security.sasl.SaslException: Invalid token in javax.security.sasl.qop: 
    at com.sun.security.sasl.util.AbstractSaslImpl.parseProp(AbstractSaslImpl.java:242)
    at com.sun.security.sasl.util.AbstractSaslImpl.parseQop(AbstractSaslImpl.java:206)
    at com.sun.security.sasl.util.AbstractSaslImpl.parseQop(AbstractSaslImpl.java:197)
    at com.sun.security.sasl.util.AbstractSaslImpl.<init>(AbstractSaslImpl.java:73)
    at com.sun.security.sasl.digest.DigestMD5Base.<init>(DigestMD5Base.java:174)
    at com.sun.security.sasl.digest.DigestMD5Server.<init>(DigestMD5Server.java:145)
    at com.sun.security.sasl.digest.FactoryImpl.createSaslServer(FactoryImpl.java:109)
    at javax.security.sasl.Sasl.createSaslServer(Sasl.java:567)
    at org.apache.hadoop.hdfs.protocol.datatransfer.sasl.SaslParticipant.createServerSaslParticipant(SaslParticipant.java:66)
    at org.apache.hadoop.hdfs.protocol.datatransfer.sasl.SaslDataTransferServer.doSaslHandshake(SaslDataTransferServer.java:385)
    at org.apache.hadoop.hdfs.protocol.datatransfer.sasl.SaslDataTransferServer.getEncryptedStreams(SaslDataTransferServer.java:186)
    at org.apache.hadoop.hdfs.protocol.datatransfer.sasl.SaslDataTransferServer.receive(SaslDataTransferServer.java:118)
    at org.apache.hadoop.hdfs.server.datanode.DataXceiver.run(DataXceiver.java:234)
    at java.lang.Thread.run(Thread.java:748)

This happens because presto still uses hadoop 2 libraries which results in failure when running on hadoop 3 system. This PR is to update support of hadoop 3 to presto engine by upgrading hadoop to version 3.2.0. This changes are not tested on hadoop 2 environment after updating it to hadoop 3 and if visibility is needed I can update that.

To add some historical context on how PrestoDB moved from hadoop-apache1 to hadoop-apache2, Presto was using hadoop-apache1 and then plugin for hadoop-apache2 was [added](https://github.com/prestodb/presto/pull/943) as well. After sometime plugin for hadoop-apache1 was [removed](https://github.com/prestodb/presto/pull/8213) from the dependencies and it started using only hadoop-apache2.
Also, Trino has moved away from hadoop 2 and is now on hadoop 3. Reference: [Update to Hadoop 3.2.0 #3](https://github.com/trinodb/trino-hadoop-apache/pull/3)

Ideally, we should also look at renaming this repository to **_presto-hadoop-apache_** to make it more generic or a new repository **_presto-hadoop-apache3_** should be created for hadoop version 3 the way repo [**_presto-hadoop-apache1_**](https://github.com/prestodb/presto-hadoop-apache1) and [**_presto-hadoop-apache2_** ](https://github.com/prestodb/presto-hadoop-apache2)already exist for version 1 and 2 respectively. Since, I am not sure which route needs to be taken, I didn't create or tried to rename the repository. Please let me know on the steps to be taken and I can follow those steps to do the further work required.

### CODE REVIEW

**Comment by rajatrj20** [comment]
@highker Could you please review this PR for me?

File: pom.xml [hunk 1]
```diff
@@ -402,7 +448,7 @@
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-shade-plugin</artifactId>
-                <version>2.4.2</version>
+                <version>3.2.1</version>
```
Comment (arunthirupathi) [inline pom.xml:451]: What was the old shaded jar size versus the new size ?

File: pom.xml [hunk 2]
```diff
@@ -426,10 +472,18 @@
                                     <pattern>org.apache.http</pattern>
                                     <shadedPattern>${shadeBase}.org.apache.http</shadedPattern>
                                 </relocation>
+                                <relocation>
+                                    <pattern>mozilla</pattern>
```
Comment (arunthirupathi) [inline pom.xml:476]: are these new dependencies from hadoop upgrade ? Seems like hadoop3 has lot of new dependencies.

**Review by arunthirupathi — COMMENTED** [review]
Can you please elaborate what features are currently broken, because of the old Hadoop version ?

File: pom.xml (hunk 1 above)
Comment (rajatrj20) [reply pom.xml:451]: This is to add support for Hadoop 3 to presto so presto works when environment has Hadoop 3. 

Currently, some queries fail when it runs on Hadoop 3. It is very likely that once we start supporting Hadoop 3 things will start failing for Hadoop 2, so the recommendation is to create separate repository to support Hadoop 3.

Comment (arunthirupathi) [reply pom.xml:451]: I am not an expert on Hadoop, so will defer this to Hadoop experts.

**Comment by rajatrj20** [comment]
@arunthirupathi @imjalpreet, Can you please help me review this PR. I have updated it with more information since last change.

**Comment by arunthirupathi** [comment]
> @arunthirupathi @imjalpreet, Can you please help me review this PR. I have updated it with more information since last change.

I work for Meta/Facebook. We don't use the hadoop at all, so not sure, if this would break anything. @imjalpreet  do you have concerns ?

My concerns are around the changed dependency and how it impacts other libraries already out there.

**Comment by imjalpreet** [comment]
@arunthirupathi @rajatrj20 Thanks for the PR. I will try to review this early next week and let you know.

**Comment by imjalpreet** [comment]
@rajatrj20 Apologies for such a long delay. I will complete the review this week. But before I begin, I have a small clarification to make. Since we are taking reference from a Trino PR https://github.com/trinodb/trino-hadoop-apache/pull/3, is there any reason behind picking only partial changes? IMO we should cherry pick the entire commit.

**Comment by rajatrj20** [comment]
@imjalpreet I decided to leave classes that will not be used in PrestoDB from the original commit. Also, there are some differences in excluded dependencies between the trino commit and this change, hence the original commit was not cherry-picked as it is.

**Comment by imjalpreet** [comment]
@rajatrj20 oh okay, I will take a look. 

Also, I see there is already a tag created for 3.2.0-1: https://github.com/prestodb/presto-hadoop-apache2/tree/3.2.0-1 which has all the required changes.

@aweisberg Do you have some context on this tag and can it be used for the upgrade?

**Comment by aweisberg** [comment]
> @rajatrj20 oh okay, I will take a look.
> 
> Also, I see there is already a tag created for 3.2.0-1: https://github.com/prestodb/presto-hadoop-apache2/tree/3.2.0-1 which has all the required changes.
> 
> @aweisberg Do you have some context on this tag and can it be used for the upgrade?

You should be able to use it. I think it was already released to maven. I don’t recall why I didn’t complete the upgrade.

**Comment by imjalpreet** [comment]
> > @rajatrj20 oh okay, I will take a look.
> > Also, I see there is already a tag created for 3.2.0-1: https://github.com/prestodb/presto-hadoop-apache2/tree/3.2.0-1 which has all the required changes.
> > @aweisberg Do you have some context on this tag and can it be used for the upgrade?
> 
> You should be able to use it. I think it was already released to maven. I don’t recall why I didn’t complete the upgrade.

@rajatrj20  I think it should be good as well, what do you think about this tag? It is already released to maven as well (https://mvnrepository.com/artifact/com.facebook.presto.hadoop/hadoop-apache2/3.2.0-1) 

> Ideally, we should also look at renaming this repository to presto-hadoop-apache to make it more generic or a new repository presto-hadoop-apache3 should be created for hadoop version 3 the way repo [presto-hadoop-apache1](https://github.com/prestodb/presto-hadoop-apache1) and [presto-hadoop-apache2 ](https://github.com/prestodb/presto-hadoop-apache2)already exist for version 1 and 2 respectively. Since, I am not sure which route needs to be taken, I didn't create or tried to rename the repository. Please let me know on the steps to be taken and I can follow those steps to do the further work required.

@aweisberg we should think about this too. Do we want to rename the repository or create a new for hadoop3?

**Comment by Ayan07** [comment]
@imjalpreet Is it possible now to build presto with presto-hadoop-apache 3 version??

**Comment by rajatrj20** [comment]
> @rajatrj20 I think it should be good as well, what do you think about this tag? It is already released to maven as well (https://mvnrepository.com/artifact/com.facebook.presto.hadoop/hadoop-apache2/3.2.0-1)

@imjalpreet What are you suggesting here - Do you want me to rebase my changes on top of 3.2.0-1 tag? Also, I am not sure if we can release same tag again to maven central please confirm if it can be done.

**Comment by aweisberg** [comment]
I can rename it, but the bigger issue is getting it merged into Presto and having a working Presto PR to pair with it.

**Comment by rajatrj20** [comment]
Closing this PR as it is now failing EasyCLA check. Opened PR against branch `3.2.x` Update hadoop to 3.2.0 #56

**Pull request closed without merging**